libnetwork/endpoint.go

```diff
@@ -595,36 +595,103 @@ func (ep *Endpoint) sbJoin(ctx context.Context, sb *Sandbox, options ...Endpoint
+			undoFunc, err := gwepBefore4.revokeExternalConnectivity()
 			if err != nil {
-				return fmt.Errorf("failed to get driver for revoking external connectivity during join: %v", err)
+				return err
+			}
+			if restartGw4 {
+				// The IPv4 gateway hasn't changed, but its noProxy6To4 setting has. So,
+				// restore it as the gateway with that new setting.
+				log.G(ctx).Debugf("Programming IPv4 gateway endpoint %s (%s) NoProxy6To4:%v",
+					ep.Name(), ep.ID(), noProxy6To4After)
+				labelsAfter := sb.Labels()
+				labelsAfter[netlabel.NoProxy6To4] = noProxy6To4After
+				if err := undoFunc(ctx, labelsAfter); err != nil {
+					log.G(ctx).WithFields(log.Fields{
+						"endpointName": ep.Name(),
+						"endpointId":   ep.ID(),
+						"error":        err,
+					}).Warn("Failed to restore IPv4 connectivity")
+				}
+			} else {
+				defer func() {
+					if retErr != nil {
+						labelsBefore := sb.Labels()
+						labelsBefore[netlabel.NoProxy6To4] = noProxy6To4Before
+						if err := undoFunc(ctx, labelsBefore); err != nil {
+							log.G(ctx).WithFields(log.Fields{
+								"endpointName": ep.Name(),
+								"endpointId":   ep.ID(),
+								"role":         role,
+								"error":        err,
+							}).Warn("Failed to restore connectivity during rollback")
+						}
+					}
+				}()
 			}
-			if err := extD.RevokeExternalConnectivity(extEp.network.ID(), extEp.ID()); err != nil {
-				return types.InternalErrorf(
-					"driver failed revoking external connectivity on endpoint %s (%s): %v",
-					extEp.Name(), extEp.ID(), err)
+		}
+		// If ep is the new IPv6 gateway, there's an old IPv6 gateway to remove, and it
+		// wasn't also the IPv4 gateway (removed already) - remove the old gateway.
+		if ep == gwepAfter6 && gwepBefore6 != nil && gwepBefore6 != gwepBefore4 {
+			log.G(ctx).Debugf("Programming IPv6 gateway endpoint %s (%s)", ep.Name(), ep.ID())
+			undoFunc, err := gwepBefore6.revokeExternalConnectivity()
+			if err != nil {
+				return err
 			}
 			defer func() {
 				if retErr != nil {
-					if e := extD.ProgramExternalConnectivity(context.WithoutCancel(ctx), extEp.network.ID(), extEp.ID(), sb.Labels()); e != nil {
-						log.G(ctx).Warnf("Failed to roll-back external connectivity on endpoint %s (%s): %v",
-							extEp.Name(), extEp.ID(), e)
+					if err := undoFunc(ctx, sb.Labels()); err != nil {
+						log.G(ctx).WithFields(log.Fields{
+							"endpointName": ep.Name(),
+							"endpointId":   ep.ID(),
+							"error":        err,
+						}).Warn("Failed to restore IPv6 connectivity during rollback")
 					}
 				}
 			}()
 		}
 		if !n.internal {
 			log.G(ctx).Debugf("Programming external connectivity on endpoint %s (%s)", ep.Name(), ep.ID())
-			if err = d.ProgramExternalConnectivity(ctx, n.ID(), ep.ID(), sb.Labels()); err != nil {
+			labels := sb.Labels()
+			labels[netlabel.NoProxy6To4] = noProxy6To4After
+			if err := d.ProgramExternalConnectivity(ctx, n.ID(), ep.ID(), labels); err != nil {
```

Comment:
Any reason not to use the newly introduced `ep.programExtrernalConnectivity` here?

Reply:
`ep.programExternalConnectivity` loads the driver and network from the store. It's used by `sbLeave` when the departing endpoint is a gateway, and some other Endpoint needs to take over the gateway-ness. That other endpoint may have a different driver, and will have a different network.

Here, during `sbJoin`, the driver and network are `ep`'s - so they're already loaded and the driver call can be made directly.